core/spec/lib/spree/core/controller_helpers/store_spec.rb

```
@@ -45,6 +45,94 @@ class FakesController < ApplicationController
     end
   end
 
+  describe '#ensure_current_store' do
+    let!(:store) { create :store, default: true }
+    let!(:store_2) { create :store }
+
+    context 'on an object that accepts multiple stores' do
+      before { allow(controller).to receive(:current_store).and_return(store) }
+
+      context 'when the object has no stores associated' do
+        object = Spree::Product.new
+
+        it 'associates the object with the current_store' do
+          controller.ensure_current_store(object)
+          expect(object.stores).to contain_exactly(store)
+          expect(object.stores).not_to contain_exactly(store_2)
+        end
+      end
+
+      context 'when the object has a store pre assigned' do
+        object = Spree::Product.new
+
+        it 'adds the new store without removing the orgional store' do
+          object.stores << store_2
+          object.save
+
+          controller.ensure_current_store(object)
+          expect(object.stores).to contain_exactly(store, store_2)
+        end
+      end
+
+      context 'when the object has a store and the same store is attempted to be added' do
+        object = Spree::Product.new
+
+        it 'object is not changed' do
+          object.stores << store
+          object.save
+
+          controller.ensure_current_store(object)
+          expect(object.stores).to contain_exactly(store)
+        end
+      end
+    end
+
+    context 'on a object that accepts a single store' do
+      before { allow(controller).to receive(:current_store).and_return(store) }
+
+      context 'when no store is present' do
+        object = Spree::Menu.new
+
+        it 'sets the current_store' do
+          controller.ensure_current_store(object)
+          expect(object.store).to eql(store)
+          expect(object.store).not_to eql(store_2)
+        end
+      end
+
+      context 'when an object already has a store assigned' do
```

@MatthewKennedy I think in this case we should raise an exception